050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,876 @@
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/node-pools: enabled
+spec:
+  kafka:
+    version: 3.3.1
+    listeners:
+      - name: plain
+        port: 9092
+        type: internal
+        tls: false
+      - name: tls
+        port: 9093
+        type: internal
+        tls: true
+    config:
+      offsets.topic.replication.factor: 3
+      transaction.state.log.replication.factor: 3
+      transaction.state.log.min.isr: 2
+      default.replication.factor: 3
+      min.insync.replicas: 2
+      inter.broker.protocol.version: "3.3"
+  # ...
+---
+
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: kafka
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  storage:
+    type: jbod
+    volumes:
+    - id: 0
+      type: persistent-claim
+      size: 1000Gi
+      class: fast-storage
+  # ...
+```
+
+Because the `KafkaNodePool` is named `kafka`, all the resources created by this configuration will be exactly the same and have the same names as today.
+Using a name different from `kafka` would cause a new Kafka nodes to be created instead of migrating the existing nodes.
+Once the initial migration is complete, users can continue to use the node pools without any limitation.
+For example add additional node pools or further modify the `kafka` node pool.
+
+#### Migration to KRaft
+
+As mentioned earlier, KRaft will be supported only with the use of the `KafkaNodePool` resources. 
+_(The detailed description of the migration to KRaft is not part of this proposal.)_
+Thanks to that, latest when ZooKeeper support is removed from Strimzi and Apache Kafka, all users will have to migrate from using the _virtual node pool_ to the `KafkaNodePools`.
+So once the ZooKeeper support is removed from Strimzi, we will also remove the support for the _virtual node pool_ and the `strimzi.io/node-pools` annotation.
+And all Strimzi based Apache Kafka clusters will have to use the `KafkaNodePool` resources
+
+## Risks
+
+### Eventual consistency
+
+Using multiple resources in Kubernetes is tricky as we have to rely on eventual consistency.
+Users can store multiple YAML documents in a single file and apply them with a single command.
+But their creation is not a single operation.
+It can therefore happen that the operator first sees only the `KafkaNodePool` resources but not the `Kafka` resource or the other way around.
+The basic situations can be handled by the operator code.
+But it is not completely clear how many problems and various race conditions this might be causing when updating the node pools for existing clusters etc.
+
+The eventual consistency also shows up in situations when you for example decide to scale two of your node pools.
+The scale events will be independent and will almost never be caught by the same reconciliation.
+So you will have one reconciliation to scale the first pool followed by the second to scale the second one.
+
```

Comment:
I guess we should be clear in this section that it's up to the users of the operator to take account of this. We should also mention that the Strimzi documentation will reflect this, for example, by procedures changing individual resources and waiting for readiness, so that the CR changes in any procedure have deterministic serialization.